### PR TITLE
Search page continuation token takes extra apiVersion and nextLink

### PR DESCRIPTION
**Pull request by sima-zhu**
Continuation token json is as follows:
```
{
    "apiVersion": "2019-05-06-Preview",
    "nextLink": "https://...",
    "nextPageParameters": {
        ...
    }
}
```
```
String continuationToken = Base64.getEncoder().encodeToString("$json");
```

### CODE REVIEW

**Comment by alzimmermsft** [comment]
@tg-msft @xirzec @xiangyan99 mind taking a quick look to ensure that all languages are using the same data structure to cover scenarios, which Ted mentioned, such as a processing queue.

**Comment by sima-zhu** [comment]
https://github.com/Azure/azure-sdk-for-net/issues/10590

File: sdk/search/azure-search-documents/pom.xml [hunk 1]
```diff
@@ -134,12 +128,6 @@
       <version>1.7.3</version> <!-- {x-version-update;com.microsoft.azure:azure-client-authentication;external_dependency} -->
       <scope>test</scope>
     </dependency>
-    <dependency>
-      <groupId>com.microsoft.azure</groupId>
```
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/search/azure-search-documents/pom.xml:138]: I have tested over. This is track 1 search lib. I am not sure where it needs the library. Live tests and playback passed.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedResponse.java [hunk 2]
```diff
@@ -27,24 +35,30 @@
  * to true coverage - coverage value.
  */
 @Immutable
-public final class SearchPagedResponse implements ContinuablePage<SearchRequest, SearchResult>,
-    Response<List<SearchResult>> {
+public final class SearchPagedResponse extends PagedResponseBase<SearchResult, SearchResult> {
```
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedResponse.java:36]: Changed to PagedResponseBase<Void, SearchResult> for convenience getter.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java [hunk 3]
```diff
@@ -643,6 +643,21 @@ private static SearchRequest createSearchRequest(String searchText, SearchOption
...
+     * Build request from continuation token.
+     *
+     * @param continuationToken The token used to deserialize the SearchRequest.
+     * @return The SearchRequest for next page.
+     */
+    private SearchRequest buildRequestFromToken(String continuationToken) {
+        try {
+            return new JacksonAdapter().deserialize(continuationToken, SearchRequest.class, SerializerEncoding.JSON);
+        } catch (IOException e) {
+            throw logger.logExceptionAsError(new RuntimeException("Failed to deserialize the token to search request"
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java:656]: Based on that I would go with `IllegalStateException` which is meant to cover scenarios where the application ran into an unrecoverable state, `IllegalArgumentException` is reserved for scenarios where a user passed in a bad value.

File: sdk/search/azure-search-documents/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/pom.xml:138]: Well that is interesting, if that is the case feel free to remove this.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tg-msft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java:656]: That is the only case we're going to get an error for though - when the user passes us a continuationToken that doesn't correctly deserialize.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java [hunk 4]
```diff
@@ -643,6 +643,21 @@ private static SearchRequest createSearchRequest(String searchText, SearchOption
...
 
+    /**
+     * Build request from continuation token.
+     *
+     * @param continuationToken The token used to deserialize the SearchRequest.
+     * @return The SearchRequest for next page.
+     */
+    private SearchRequest buildRequestFromToken(String continuationToken) {
+        try {
+            return new JacksonAdapter().deserialize(continuationToken, SearchRequest.class, SerializerEncoding.JSON);
```
Comment (tg-msft) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java:654]: I'm not sure this is correct - doesn't the continuationToken include the `apiVersion` and `nextLink`?  I think we'd only do this for the `nextPageParameters` property.  We also need to verify that this has an `apiVersion` we know how to parse and throw an exception if it has an API Version we don't recognize.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedResponse.java [hunk 5]
```diff
@@ -103,33 +127,13 @@ public Long getCount() {
...
     @Override
-    public IterableStream<SearchResult> getElements() {
-        return new IterableStream<>(value);
-    }
-
-    @Override
-    public SearchRequest getContinuationToken() {
-        return nextPageParameters;
+    public String getContinuationToken() {
+        return nextParameters;
```
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedResponse.java:137]: Will make this return the encoded string.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedResponse.java [hunk 6]
```diff
@@ -27,45 +33,63 @@
...
-    private final HttpHeaders headers;
-    private final HttpRequest request;
+public final class SearchPagedResponse extends PagedResponseBase<Void, SearchResult> {
     private final List<SearchResult> value;
 
-    private final SearchRequest nextPageParameters;
     private final Map<String, List<FacetResult>> facets;
     private final Long count;
     private final Double coverage;
+    private final String nextParameters;
```
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedResponse.java:42]: Remove nextParameters getter and use getContinuationToken return encoded token instead.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedResponse.java [hunk 7]
```diff
@@ -27,45 +33,63 @@
...
             || result.getNextPageParameters().getSkip() == null) {
             return null;
         }
-
-        return result.getNextPageParameters();
+        try {
+            return new JacksonAdapter().serialize(result.getNextPageParameters(), SerializerEncoding.JSON);
+        } catch (IOException ex) {
+            throw new RuntimeException("Failed to serialize the search request.");
+        }
```
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedResponse.java:92]: Yes, will make the change.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java (hunk 4 above)
Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java:654]: You are right when we have the getContinuationToken return encoded one, it includes more items. I will fix the code here.

Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java:654]: Will do apiVersion check as well.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sima-zhu) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchIndexAsyncClient.java]: Thanks for the discussion! 
nextLink and nextParameter are both provided by service.
apiVersion is sdk library enum. 

These three things all from SDK and above, and user has no way to manipulate it.
I will go with `IllegalStateException`

**Pull request merged**